include/jemalloc/internal/ql.h

```diff
@@ -3,69 +3,119 @@
 
 #include "jemalloc/internal/qr.h"
 
+/*
+ * A linked-list implementation.
+ *
+ * This is built on top of the ring implementation, but that can be viewed as an
+ * implementation detail (i.e. trying to advance past the tail of the list
+ * doesn't wrap around).
+ *
+ * You define a struct like so:
+ * typedef strucy my_s my_t;
+ * struct my_s {
+ *   int data;
+ *   ql_elm(my_t) my_link;
+ * };
+ *
+ * // We wobble between "list" and "head" for this type; we're now mostly
+ * // heading towards "list".
+ * typedef ql_head(my_t) my_list_t;
+ *
+ * You then pass a my_list_t * for a_head arguments, a my_t * for a_elm
+ * arguments, the token "my_link" for a_field arguments, and the token "my_t"
+ * for a_type arguments.
+ */
+
 /* List definitions. */
 #define ql_head(a_type)							\
 struct {								\
 	a_type *qlh_first;						\
 }
 
+/* Static initializer for an empty list. */
 #define ql_head_initializer(a_head) {NULL}
 
+/* The field definition. */
 #define ql_elm(a_type)	qr(a_type)
 
-/* List functions. */
+/* A pointer to the first element in the list, or NULL if the list is empty. */
 #define ql_first(a_head) ((a_head)->qlh_first)
 
+/* Dynamically initializes a list. */
 #define ql_new(a_head) do {						\
 	ql_first(a_head) = NULL;					\
 } while (0)
 
+/* Removes all elements from the list. */
 #define ql_clear(a_head) ql_new(a_head)
```

Comment:
This is something I thought about earlier but didn't have a clear conclusion. This `ql_clear()` is mainly meant to be used internally (within this header), and "responsibly", in the sense that the content of the list has always been safely relocated elsewhere before it's called, rather than risking being lost if the caller is not careful. Perhaps changing the name to `ql_clear_internal()` or `_ql_clear()` makes more sense in this case. We can also expect the caller to be responsible themselves and state this in the comment, but I couldn't find an intuitive example why the caller may want to clear the list first and then manipulate the nodes themselves. I think I'm now slightly leaning towards keeping it internal, but either way works.

Reply:
TBH I think the fact that we have both `ql_clear` and `ql_new` is an oversight. You do need to clear nonempty lists sometimes (e.g. arena tcache list after forking), but I don't know that that was the intent (and practically, we use ql_new there).

Reply:
Put up a version that just only has ql_new.

Reply:
I originally put `ql_new()`s in these places but felt that the name sounded a bit "intrusive", because we're not creating new lists there, so I added the `ql_clear()` as a "rephrase". Now it seems `ql_new()` feels better to me. I think it's fine.